src/CastMsg.chpl

```diff
@@ -17,174 +17,340 @@ module CastMsg {
-            }
-            when (DType.Bool, "int64") {
-                return new MsgTuple(castGenSymEntry(gse, st, bool, int), MsgType.NORMAL);
-            }
-            when (DType.Bool, "uint8") {
-                return new MsgTuple(castGenSymEntry(gse, st, bool, uint(8)), MsgType.NORMAL);
-            }
-            when (DType.Bool, "uint64") {
-                return new MsgTuple(castGenSymEntry(gse, st, bool, uint), MsgType.NORMAL);
-            }
-            when (DType.Bool, "float64") {
-                return new MsgTuple(castGenSymEntry(gse, st, bool, real), MsgType.NORMAL);
-            }
-            when (DType.Bool, "bool") {
-                return new MsgTuple(castGenSymEntry(gse, st, bool, bool), MsgType.NORMAL);
-            }
-            when (DType.Bool, "str") {
-                return new MsgTuple(castGenSymEntryToString(gse, st, bool), MsgType.NORMAL);
-            }
-            when (DType.Bool, "bigint") {
-                return new MsgTuple(castGenSymEntryToBigInt(gse, st, bool), MsgType.NORMAL);
-            }
-            when (DType.BigInt, "bigint") {
-                return new MsgTuple(castGenSymEntryToBigInt(gse, st, bigint), MsgType.NORMAL);
-            }
-            when (DType.BigInt, "uint64") {
-                return new MsgTuple(castGenSymEntry(gse, st, bigint, uint), MsgType.NORMAL);
-            }
-            when (DType.BigInt, "int64") {
-                return new MsgTuple(castGenSymEntry(gse, st, bigint, int), MsgType.NORMAL);
-            }
-            when (DType.BigInt, "str") {
-                return new MsgTuple(castGenSymEntryToString(gse, st, bigint), MsgType.NORMAL);
-            }
+            when (DType.Int64, DType.Int8) do return doScalarCast(int, int(8));
+            when (DType.Int64, DType.Int16) do return doScalarCast(int, int(16));
+            when (DType.Int64, DType.Int32) do return doScalarCast(int, int(32));
+            when (DType.Int64, DType.Int64) do return doScalarCast(int, int);
+            when (DType.Int64, DType.UInt8) do return doScalarCast(int, uint(8));
+            when (DType.Int64, DType.UInt16) do return doScalarCast(int, uint(16));
+            when (DType.Int64, DType.UInt32) do return doScalarCast(int, uint(32));
+            when (DType.Int64, DType.UInt64) do return doScalarCast(int, uint);
+            when (DType.Int64, DType.Float32) do return doScalarCast(int, real(32));
+            when (DType.Int64, DType.Float64) do return doScalarCast(int, real(64));
+            when (DType.Int64, DType.Complex64) do return doScalarCast(int, complex(64));
+            when (DType.Int64, DType.Complex128) do return doScalarCast(int, complex(128));
+            when (DType.Int64, DType.Bool) do return doScalarCast(int, bool);
+            when (DType.Int64, DType.Strings) do return doStringCast(int);
+            when (DType.Int64, DType.BigInt) do return doBigintCast(int);
```

Comment:
I think a macro of some kind would be very nice here because this pattern is gonna be repeated across a bunch of the message handlers once the array-api is finished. 

One way I've thought of to make that happen is an annotation + a python script that transforms the code before it's built.  Maybe something like the following could get transformed into the above code:
```chapel

@arkouda.permuteTypeInstantiation(gse.dtype, targetDtype)
doScalarCast(AllScalarTypes, AllScalarTypes);

@arkouda.permuteTypeInstantiation(gse.dtype, targetDtype == DType.Strings)
doStringCast(AllScalarTypes)

@arkouda.permuteTypeInstantiation(gse.dtype, targetDtype == DType.BigInt)
doBigintCast(AllScalarTypes)
```

That might be a heavy lift (and might make the code less maintainable?), so I'm thinking about adding it in the future.